### PR TITLE
Support Redis instrumentation for keyed service

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1451.

## Changes

Added a new overload to `AddRedisInstrumentation` which gets `serviceKey` in input and get `IConnectionMultiplexer` with a keyed service.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #
* [x] Changes in public API reviewed

I'm not sure if I add the reference to the `Microsoft.Extensions.DependencyInjection.Abstractions` correctly and Also I have problem in one of the tests which needs `ServiceCollection`. With this new package reference `ServiceCollection` exists in both 3.1.0 and 8.0.0.

### CODE REVIEW

**Comment by Kielek** [comment]
@CodeBlanch, please check.

Ref: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1451#issuecomment-1832457720

File: test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/StackExchangeRedisCallsInstrumentationTests.cs [hunk 1]
```diff
@@ -317,37 +328,37 @@ public async Task ProfilerSessionsHandleMultipleSpans()
...
-        services.AddOpenTelemetry().WithTracing(builder => builder.AddRedisInstrumentation());
-
-        using var serviceProvider = services.BuildServiceProvider();
-
-        var tracerProvider = serviceProvider.GetRequiredService<TracerProvider>();
-
-        Assert.True(connectionMultiplexerPickedFromDI);
-        Assert.True(optionsPickedFromDI);
-    }
+    //[Fact]
```
Comment (Kielek) [inline test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/StackExchangeRedisCallsInstrumentationTests.cs:331]: Was it intentional?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj [hunk 2]
```diff
@@ -19,7 +19,11 @@
     <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="$(OpenTelemetryCoreLatestVersion)" />
     <PackageReference Include="StackExchange.Redis" Version="$(StackExchangeRedisPkgVer)" />
     <PackageReference Include="Microsoft.Extensions.Options" Version="$(MicrosoftExtensionsOptionsPkgVer)" />
-    <PackageReference Include="System.Reflection.Emit.Lightweight" Version="$(SystemReflectionEmitLightweightPkgVer)" />
+    <PackageReference Include="System.Reflection.Emit.Lightweight" Version="$(SystemReflectionEmitLightweightPkgVer)" />    
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="8.0.0" />
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:26]: Any option to temporary switch version of `OpenTelemetry.Api.ProviderBuilderExtensions` to `OpenTelemetryCoreLatestPrereleaseVersion`?

Wait for 1.7.0 GA release. Revert this change and make a release of this package?

**Review by Kielek — COMMENTED** [review]
Please add entry to the CHANGELOG.md

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 3]
```diff
@@ -111,6 +111,30 @@ public static class TracerProviderBuilderExtensions
...
+    /// <remarks>
+    /// Note: If an <see cref="IConnectionMultiplexer"/> is not supplied
+    /// using the <paramref name="connection"/> parameter it will be
+    /// resolved using the application <see cref="IServiceProvider"/>.
+    /// </remarks>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
+    /// <param name="name">Optional name which is used when retrieving options.</param>
+    /// <param name="connection">Optional <see cref="IConnectionMultiplexer"/> to instrument.</param>
+    /// <param name="configure">Optional callback to configure options.</param>
+    /// <param name="serviceKey">The service key to get <see cref="IConnectionMultiplexer"/>.</param>
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:130]:
```suggestion
    /// <param name="serviceKey">Optional service key used to retrieve the <see cref="IConnectionMultiplexer"/> to instrument from the <see cref="IServiceProvider" />.</param>
```

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 4]
```diff
@@ -90,7 +90,7 @@ public static class TracerProviderBuilderExtensions
         Guard.ThrowIfNull(connection);
         Guard.ThrowIfNull(configure);
 
-        return AddRedisInstrumentation(builder, name: null, connection, configure);
+        return AddRedisInstrumentation(builder, name: null, connection, configure, null);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:93]:
```suggestion
        return AddRedisInstrumentation(builder, name: null, connection, configure, serviceKey: null);
```

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:26]: I think we should do...

```diff
-   <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="$(OpenTelemetryCoreLatestVersion)" />
+   <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="$(OpenTelemetryCoreLatestPrereleaseVersion)" />
```

Because typically some maintainer will just come and flip that for the whole repo once stable is released.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 5]
```diff
@@ -111,6 +111,30 @@ public static class TracerProviderBuilderExtensions
...
+    }
+
+    /// <summary>
+    /// Enables automatic data collection of outgoing requests to Redis.
+    /// </summary>
+    /// <remarks>
+    /// Note: If an <see cref="IConnectionMultiplexer"/> is not supplied
+    /// using the <paramref name="connection"/> parameter it will be
+    /// resolved using the application <see cref="IServiceProvider"/>.
+    /// </remarks>
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:125]: nit: We now have this duplicated with the method above. What I would suggest is use `inheritdoc` to share it.

To modify the one above to reference the new uber one which has all the parameters switch line 99 to be:

`/// <remarks><inheritdoc cref="AddRedisInstrumentation(TracerProviderBuilder, string?, IConnectionMultiplexer?, Action{StackExchangeRedisInstrumentationOptions}?, object?)" path="/remarks"/></remarks>`

Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:125]: Actually on second thought, I think we could just modify the existing uber one. That would be a breaking change but the project is still prerelease so that is OK to do. No one should really be calling the uber one, it is just there for named options which is rare. I suggest these modifications...

```diff
public static class TracerProviderBuilderExtensions
{
+    public static TracerProviderBuilder AddRedisInstrumentation(
+        this TracerProviderBuilder builder,
+        object serviceKey)
+    {
+        Guard.ThrowIfNull(serviceKey);
+
+        return AddRedisInstrumentation(builder, name: null, connection: null, configure: null, serviceKey);
+    }

+    public static TracerProviderBuilder AddRedisInstrumentation(
+        this TracerProviderBuilder builder,
+        object serviceKey,
+        Action<StackExchangeRedisInstrumentationOptions> configure)
+    {
+        Guard.ThrowIfNull(serviceKey);
+        Guard.ThrowIfNull(configure);
+
+        return AddRedisInstrumentation(builder, name: null, connection: null, configure, serviceKey);
+    }

    public static TracerProviderBuilder AddRedisInstrumentation(
        this TracerProviderBuilder builder,
        string? name,
        IConnectionMultiplexer? connection,
        Action<StackExchangeRedisInstrumentationOptions>? configure,
+       object? serviceKey)
+    {
+         // Implementation not shown
+    }
}
```

The idea is to add overloads for the DI case (where `IConnectionMultiplexer` is not passed directly) which allow passing `serviceKey`.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 6]
```diff
@@ -132,7 +156,18 @@ public static class TracerProviderBuilderExtensions
             {
                 var instrumentation = sp.GetRequiredService<StackExchangeRedisInstrumentation>();
 
-                connection ??= sp.GetService<IConnectionMultiplexer>();
+                if (connection is not null)
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:159]:
```suggestion
                if (connection is null)
```

?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 7]
```diff
@@ -132,7 +156,18 @@ public static class TracerProviderBuilderExtensions
...
+                {
+                    if (serviceKey is null)
+                    {
+                        connection = sp.GetService<IConnectionMultiplexer>();
+                    }
+                    else
+                    {
+                        connection = sp.GetKeyedService<IConnectionMultiplexer>(serviceKey);
+                    }
+                }
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:169]:
```suggestion
                connection ??= serviceKey is null ?
                    sp.GetService<IConnectionMultiplexer>() :
                    sp.GetKeyedService<IConnectionMultiplexer>(serviceKey);
```

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 8]
```diff
@@ -132,7 +156,18 @@ public static class TracerProviderBuilderExtensions
...
+
 
                 if (connection != null)
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:172]:
```suggestion
                if (connection is not null)
```

Might as well do the null checks in the same method in a consistent way.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 9]
```diff
@@ -132,8 +167,15 @@ public static class TracerProviderBuilderExtensions
...
-                connection ??= sp.GetService<IConnectionMultiplexer>();
-
+#if NET8_0_OR_GREATER
+                connection ??= serviceKey == null
+                    ? sp.GetService<IConnectionMultiplexer>()
+                    : sp.GetKeyedService<IConnectionMultiplexer>(serviceKey);
+#else
+                connection ??= serviceKey == null
+                    ? sp.GetService<IConnectionMultiplexer>()
+                    : throw new InvalidOperationException();
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:177]: Should this have a message?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 10]
```diff
@@ -53,7 +53,40 @@ public static class TracerProviderBuilderExtensions
...
+    /// Enables automatic data collection of outgoing requests to Redis.
+    /// </summary>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
+    /// <param name="serviceKey">Optional service key used to retrieve the <see cref="IConnectionMultiplexer"/> to instrument from the <see cref="IServiceProvider" />.</param>
+    /// <param name="configure">Callback to configure options.</param>
+    /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+    public static TracerProviderBuilder AddRedisInstrumentation(
+        this TracerProviderBuilder builder,
+        object serviceKey,
+        Action<StackExchangeRedisInstrumentationOptions> configure)
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:84]: Do we need a way to specify the `name`? As it stands, if I use this API the `configure` callback will configure the `StackExchangeRedisInstrumentationOptions` for all connections. What if I only want to configure the options for this specific connection?

**Comment by eerhardt** [comment]
@Kahbazi - any chance you can sync the branch and address the 2 comments above?

**Comment by Kielek** [comment]
@Kahbazi, could you please check leftovers?

**Comment by Kahbazi** [comment]
@Kielek Sorry for the delay. I will finish this PR this week.

**Comment by eerhardt** [comment]
@Kahbazi - it looks like some CI checks are failing. Can you take a look?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 11]
```diff
@@ -92,12 +127,14 @@ public static class TracerProviderBuilderExtensions
...
     /// <param name="configure">Optional callback to configure options.</param>
+    /// <param name="serviceKey">Optional service key used to retrieve the <see cref="IConnectionMultiplexer"/> to instrument from the <see cref="IServiceProvider" />.</param>
     /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
     public static TracerProviderBuilder AddRedisInstrumentation(
         this TracerProviderBuilder builder,
         string? name,
         IConnectionMultiplexer? connection,
-        Action<StackExchangeRedisInstrumentationOptions>? configure)
+        Action<StackExchangeRedisInstrumentationOptions>? configure,
+        object? serviceKey)
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:137]: The order of these parameters is inconsistent with the overload added above. Above we have:

* `builder`, `name`, `serviceKey`, `configure`

But here we have `serviceKey` after `configure`. I think it would make sense to have `serviceKey` before `configre` on both of them.
